gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/x509/X509AuthenticationUserDetailsService.groovy

```diff
@@ -56,7 +59,14 @@ class X509AuthenticationUserDetailsService implements AuthenticationUserDetailsS
 
     def x509 = (X509Certificate) token.credentials
     def email = emailFromSubjectAlternativeName(x509) ?: token.principal
-    def roles = userRolesProvider.loadRoles(email as String)
+
+    def roles
+    if (rolesExtractor) {
+      roles = rolesExtractor.fromCertificate(x509).orElse(userRolesProvider.loadRoles(email as String))
```

Comment:
I suggest merging the lists of roles. Roles from a cert don't necessarily have to be the complete list.